src/common/gpio_servo.rs

```diff
@@ -0,0 +1,191 @@
+    if board.is_none() {
+        return Err(anyhow::anyhow!(
+            "actual board is required to be passed to configure a gpio servo"
+        ));
+    }
+    let board_unwrapped = board.unwrap();
+    let servo_settings = GpioServoSettings::from_config(&cfg)?;
+    let pin = cfg.get_attribute::<i32>("pin").map_err(|err| match err {
+        AttributeError::ConversionImpossibleError => {
+            anyhow::anyhow!("could not convert pin attribute format when configuring gpio servo")
+        }
+        _ => anyhow::anyhow!("error configuring gpio servo: {:?}", err),
+    })?;
+    Ok(Arc::new(Mutex::new(GpioServo::<BoardType>::new(
+        board_unwrapped,
+        pin,
+        servo_settings,
+    )?)))
+}
+
+#[derive(Debug)]
+pub struct GpioServoSettings {
+    pub min_angle_deg: f64,
+    pub max_angle_deg: f64,
+    pub min_period_us: u32,
+    pub max_period_us: u32,
+    pub frequency: u32,
+    /// when 0, pwm_resolution is not considered when calculating the PWM duty cycle
+    /// necessary to move the servo to particular angular position
+    pub pwm_resolution: u32,
+}
+
+impl GpioServoSettings {
+    pub fn from_config(cfg: &ConfigType) -> anyhow::Result<Self> {
+        let min_angle_deg = cfg
+            .get_attribute::<f64>("min_angle_deg")
+            .unwrap_or_default();
+        let max_angle_deg = cfg.get_attribute::<f64>("max_angle_deg").unwrap_or(180.0);
+        let min_period_us = cfg.get_attribute::<u32>("min_width_us").unwrap_or(500);
+        let max_period_us = cfg.get_attribute::<u32>("max_period_us").unwrap_or(2500);
+        let frequency = cfg.get_attribute::<u32>("frequency_hz").unwrap_or(300);
+        let pwm_resolution = cfg
+            .get_attribute::<u32>("pwm_resolution")
+            .unwrap_or_default();
+        Ok(Self {
+            min_angle_deg,
+            max_angle_deg,
+            min_period_us,
+            max_period_us,
+            frequency,
+            pwm_resolution,
+        })
+    }
+}
+
+pub struct GpioServo<B> {
+    board: B,
+    pin: i32,
+    min_angle_deg: f64,
+    max_angle_deg: f64,
+    min_period_us: u32,
+    max_period_us: u32,
+    frequency: u32,
+    pwm_resolution: u32,
+}
+
+impl<B> GpioServo<B>
+where
+    B: Board,
+{
+    pub(crate) fn new(board: B, pin: i32, settings: GpioServoSettings) -> anyhow::Result<Self> {
+        let mut res = Self {
+            board,
+            pin,
+            min_angle_deg: settings.min_angle_deg,
+            max_angle_deg: settings.max_angle_deg,
+            min_period_us: settings.min_period_us,
+            max_period_us: settings.max_period_us,
+            frequency: settings.frequency,
+            pwm_resolution: settings.pwm_resolution,
+        };
+        res.board.set_pwm_frequency(pin, res.frequency as u64)?;
+        Ok(res)
+    }
+
+    fn angle_to_duty_pct(&self, angle_deg: u32) -> f64 {
+        let period = 1.0 / (self.frequency as f64);
```

Comment:
what happens if frequency is 0? I don't think I saw anything above that prevents that.

Reply:
@gvaradarajan - I'm interested in the answer to this.

Reply:
I'll raise an error in the `new` function if the passed frequency is 0. I think past that, it's ok if this panics because someone would really have to go out of their way to make a `GpioServo` instance with a zero frequency